core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java

```diff
@@ -130,6 +130,18 @@ public static boolean isOptionSchema(Schema schema) {
     return false;
   }
 
+  public static boolean isNonOptionalUnionOptional(Schema schema) {
+    if (schema.getType() == UNION && schema.getTypes().size() != 2) {
```

Comment:
good catch @rzhang10 ! While it is meaningless to have a single-type in a union, the avro `Schema.UnionSchema` constructor does not perform any checks on the number of types, so we definitely need to guard against such cases (throw if size < 2 ?).

Reply:
Yeah, it can happen. For example, `[null]`  and `[int]` will pass the check. However, those are valid Avro unions and Avro does not prevent users from creating those schemas. In addition, our reader will works with those single unions. Maybe it is better to support those than guard it? We can let the user to decide whether it is useful or not.